lib/defender/aws_iot_defender.c

```diff
@@ -86,19 +93,26 @@ static AwsIotTimer_t _metricsPublishTimer;
 /* the timer that disconnect */
 static AwsIotTimer_t _guardTimer;
 
+/*
+ * This semaphore is to synchronize AwsIotDefender_Stop and metrics timer callback + MQTT callback + guard timer callback.
+ * AwsIotDefender_Stop waits on this semaphore.
+ * Metrics timer callback tries to wait on this semaphore.
+ */
+static AwsIotSemaphore_t _timerSem;
```

Comment:
Do we actually need a semaphore and a mutex?

Reply:
I think the semaphore is necessary because you need to implement this logic: destroy resources only after the timer callback completes execution.

The mutex for metrics can be left out if we don't need to support modifying metrics after defender is started.